File: LeetCode/3577.cpp
// https://leetcode.com/problems/count-the-number-of-computer-unlocking-permutations/

#include <iostream>
#include <vector>

using namespace std;

int countPermutations(vector<int> &complexity)
{
    long long t = 1;
    int n = complexity.size();

    for (int i = 1; i < n; i++)
    {
        if (complexity[i] <= complexity[0])
        {
            return 0;
        }
    }

    for (int i = 1; i < n; i++)
    {
        t = (t * i) % (1000000007);
    }

    return t; // (0 ms)
}

File: LeetCode/3577.py
# https://leetcode.com/problems/count-the-number-of-computer-unlocking-permutations/

def countPermutations(complexity: list[int]) -> int:
    t = 1
    n = len(complexity)
    p = [set() for i in complexity]
    
    for i in range(1, n):
        for j in range(i):
            if complexity[j] < complexity[i]:
                p[i].add(j)
        
    u = {0}
    r = set()
    
    for i in range(1, n):
        if 0 in p[i]:
            r.add(i)
            
    for i in range(1, n):
        c = 1
        
        t = (t * c) % (10 ** 9 + 7)
        
        k = next(iter(r))
        r.remove(k)
        u.add(k)
        
    return t